hdf5_serialiser/__init__.py

```diff
@@ -201,12 +110,90 @@ def serialisable(input_class):
+    class Serialised(AbstractSerialisable, input_class):
+        def serialise(
+            self, output: FileType | h5py.File | h5py.Group
+        ):
+            def serialise_single(
+                attr: str, val: Any, T: type, h5: h5py.File | h5py.Group
+            ) -> None:
+                if val is None:
+                    return
+
+                if _is_primitive(T) or _is_optional_primitive(T):
+                    h5.attrs[attr] = val
+                elif _is_pydantic_model(T):
+                    
+                    h5.attrs[attr] = val.json()
+                # TODO: elif list - json?
+                elif _is_numpy_array(T):
+                    h5.create_dataset(attr, data=val)
+                elif _is_class_serialisable(T):
+                    grp = h5.create_group(attr)
+                    val.serialise(output=grp)
+                elif _is_supported_dict(T):
+                    _, V = get_args(T)
+                    grp = h5.create_group(attr)
+                    for k, v in val.items():
+                        serialise_single(k, v, V, grp)
+                else:
+                    raise Exception(f"Unsupported type of attribute '{attr}'")
 
-        @classmethod
-        def deserialise(cls, input: FileType | h5py.File | h5py.Group):
-            return _deserialise(cls, input, serialisable_attrs)
+            h5 = (
+                output
+                if isinstance(output, (h5py.File, h5py.Group))
+                else h5py.File(output, "w")
+            )
+
+            for attr, T in serialisable_attrs.items():
+                val = getattr(self, attr)
+                serialise_single(attr, val, T, h5)
 
-    return NewCls
+        @classmethod
+        def deserialise(
+            cls,
+            input: FileType | h5py.File | h5py.Group,
+            serialisable_attrs: dict[str, type],
+        ):
+            def deserialise_single(attr: str, T: type, h5: h5py.File | h5py.Group) -> Any:
+                val = None
+                if _is_primitive(T):
+                    val = h5.attrs.get(attr)
+                    assert (
+                        val is not None
+                    ), f"Attribute '{attr}' marked as non-optional, but value is not present!"
+                elif _is_optional_primitive(T):
+                    val = h5.attrs.get(attr)
+                elif _is_pydantic_model(T):
+                    assert issubclass(T, BaseModel)
+                    val = T.parse_raw(h5.attrs.get(attr))
+                else:
+                    serialised = h5[attr]
+                    if isinstance(serialised, h5py.Dataset):
+                        val = np.array(serialised)
+                    elif isinstance(serialised, h5py.Group):
+                        assert _is_class_serialisable(T) or _is_supported_dict(T)
+                        if is_dataclass(T) and issubclass(T, AbstractSerialisable): # _is_class_serialisable
```

Comment:
Why not just _is_class_serialisable?

Reply:
Ah this is a class type checker issue - moving it into the lower level function means it doesn't know it's a subclass of x and raises an error - not sure type predicates are a thing in python

Reply:
Replacing _is_class_serialisable(T) -> bool to _is_class_serialisable(T) -> TypeGuard[Type[AbstractSerialisable]] solves the problem.

Reply:
Huh, that's cool!